File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-msg-5-4/Reason.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Reason : ISerialized
    {
        public Reason(XmlNode xmlNode)
        {
            XmlNode reasonCodeNode = xmlNode.SelectSingleNode("reasonCode");
            
            if (reasonCodeNode != null)
            {
                if (reasonCodeNode.Attributes["href"] != null || reasonCodeNode.Attributes["id"] != null) 
                {
                    if (reasonCodeNode.Attributes["id"] != null) 
                    {
                        reasonCodeIDRef_ = reasonCodeNode.Attributes["id"].Value;
                        ReasonCode ob = new ReasonCode(reasonCodeNode);
                        IDManager.SetID(reasonCodeIDRef_, ob);
                    }
                    else if (reasonCodeNode.Attributes["href"] != null)
                    {
                        reasonCodeIDRef_ = reasonCodeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        reasonCode_ = new ReasonCode(reasonCodeNode);
                    }
                }
                else
                {
                    reasonCode_ = new ReasonCode(reasonCodeNode);
                }
            }
            
        
            XmlNode locationNode = xmlNode.SelectSingleNode("location");
            
            if (locationNode != null)
            {
                if (locationNode.Attributes["href"] != null || locationNode.Attributes["id"] != null) 
                {
                    if (locationNode.Attributes["id"] != null) 
                    {
                        locationIDRef_ = locationNode.Attributes["id"].Value;
                        ProblemLocation ob = new ProblemLocation(locationNode);
                        IDManager.SetID(locationIDRef_, ob);
                    }
                    else if (locationNode.Attributes["href"] != null)
                    {
                        locationIDRef_ = locationNode.Attributes["href"].Value;
                    }
                    else
                    {
                        location_ = new ProblemLocation(locationNode);
                    }
                }
                else
                {
                    location_ = new ProblemLocation(locationNode);
                }
            }
            
        
            XmlNode descriptionNode = xmlNode.SelectSingleNode("description");
            
            if (descriptionNode != null)
            {
                if (descriptionNode.Attributes["href"] != null || descriptionNode.Attributes["id"] != null) 
                {
                    if (descriptionNode.Attributes["id"] != null) 
                    {
                        descriptionIDRef_ = descriptionNode.Attributes["id"].Value;
                        XsdTypeString ob = new XsdTypeString(descriptionNode);
                        IDManager.SetID(descriptionIDRef_, ob);
                    }
                    else if (descriptionNode.Attributes["href"] != null)
                    {
                        descriptionIDRef_ = descriptionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        description_ = new XsdTypeString(descriptionNode);
                    }
                }
                else
                {
                    description_ = new XsdTypeString(descriptionNode);
                }
            }
            
        
            XmlNode validationRuleIdNode = xmlNode.SelectSingleNode("validationRuleId");
            
            if (validationRuleIdNode != null)
            {
                if (validationRuleIdNode.Attributes["href"] != null || validationRuleIdNode.Attributes["id"] != null) 
                {
                    if (validationRuleIdNode.Attributes["id"] != null) 
                    {
                        validationRuleIdIDRef_ = validationRuleIdNode.Attributes["id"].Value;
                        Validation ob = new Validation(validationRuleIdNode);
                        IDManager.SetID(validationRuleIdIDRef_, ob);
                    }
                    else if (validationRuleIdNode.Attributes["href"] != null)
                    {
                        validationRuleIdIDRef_ = validationRuleIdNode.Attributes["href"].Value;
                    }
                    else
                    {
                        validationRuleId_ = new Validation(validationRuleIdNode);
                    }
                }
                else
                {
                    validationRuleId_ = new Validation(validationRuleIdNode);
                }
            }
            
        
            XmlNodeList additionalDataNodeList = xmlNode.SelectNodes("additionalData");
            
            if (additionalDataNodeList != null)
            {
                this.additionalData_ = new List<AdditionalData>();
                foreach (XmlNode item in additionalDataNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            additionalDataIDRef_ = item.Attributes["id"].Value;
                            additionalData_.Add(new AdditionalData(item));
                            IDManager.SetID(additionalDataIDRef_, additionalData_[additionalData_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            additionalDataIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        additionalData_.Add(new AdditionalData(item));
                        }
                    }
                    else
                    {
                        additionalData_.Add(new AdditionalData(item));
                    }
                }
            }
            
        
        }
        
    
        #region ReasonCode_
        private ReasonCode reasonCode_;
        public ReasonCode ReasonCode_
        {
            get
            {
                if (this.reasonCode_ != null)
                {
                    return this.reasonCode_; 
                }
                else if (this.reasonCodeIDRef_ != null)
                {
                    reasonCode_ = IDManager.getID(reasonCodeIDRef_) as ReasonCode;
                    return this.reasonCode_; 
                }
                else
                {
                      return this.reasonCode_; 
                }
            }
            set
            {
                if (this.reasonCode_ != value)
                {
                    this.reasonCode_ = value;
                }
            }
        }
        #endregion
        
        public string reasonCodeIDRef_ { get; set; }
        #region Location_
        private ProblemLocation location_;
        public ProblemLocation Location_
        {
            get
            {
                if (this.location_ != null)
                {
                    return this.location_; 
                }
                else if (this.locationIDRef_ != null)
                {
                    location_ = IDManager.getID(locationIDRef_) as ProblemLocation;
                    return this.location_; 
                }
                else
                {
                      return this.location_; 
                }
            }
            set
            {
                if (this.location_ != value)
                {
                    this.location_ = value;
                }
            }
        }
        #endregion
        
        public string locationIDRef_ { get; set; }
        #region Description_
        private XsdTypeString description_;
        public XsdTypeString Description_
        {
            get
            {
                if (this.description_ != null)
                {
                    return this.description_; 
                }
                else if (this.descriptionIDRef_ != null)
                {
                    description_ = IDManager.getID(descriptionIDRef_) as XsdTypeString;
                    return this.description_; 
                }
                else
                {
                      return this.description_; 
                }
            }
            set
            {
                if (this.description_ != value)
                {
                    this.description_ = value;
                }
            }
        }
        #endregion
        
        public string descriptionIDRef_ { get; set; }
        #region ValidationRuleId_
        private Validation validationRuleId_;
        public Validation ValidationRuleId_
        {
            get
            {
                if (this.validationRuleId_ != null)
                {
                    return this.validationRuleId_; 
                }
                else if (this.validationRuleIdIDRef_ != null)
                {
                    validationRuleId_ = IDManager.getID(validationRuleIdIDRef_) as Validation;
                    return this.validationRuleId_; 
                }
                else
                {
                      return this.validationRuleId_; 
                }
            }
            set
            {
                if (this.validationRuleId_ != value)
                {
                    this.validationRuleId_ = value;
                }
            }
        }
        #endregion
        
        public string validationRuleIdIDRef_ { get; set; }
        #region AdditionalData_
        private List<AdditionalData> additionalData_;
        public List<AdditionalData> AdditionalData_
        {
            get
            {
                if (this.additionalData_ != null)
                {
                    return this.additionalData_; 
                }
                else if (this.additionalDataIDRef_ != null)
                {
                    return this.additionalData_; 
                }
                else
                {
                      return this.additionalData_; 
                }
            }
            set
            {
                if (this.additionalData_ != value)
                {
                    this.additionalData_ = value;
                }
            }
        }
        #endregion
        
        public string additionalDataIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-msg-5-4/Reason.cpp
// Reason.cpp 
#include "Reason.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Reason::Reason(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //reasonCodeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* reasonCodeNode = xmlNode->FirstChildElement("reasonCode");

   if(reasonCodeNode){reasonCodeIsNull_ = false;}
   else{reasonCodeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- reasonCodeNode , address : " << reasonCodeNode << std::endl;
   #endif
   if(reasonCodeNode)
   {
      if(reasonCodeNode->Attribute("href") || reasonCodeNode->Attribute("id"))
      {
          if(reasonCodeNode->Attribute("id"))
          {
             reasonCodeIDRef_ = reasonCodeNode->Attribute("id");
             reasonCode_ = boost::shared_ptr<ReasonCode>(new ReasonCode(reasonCodeNode));
             reasonCode_->setID(reasonCodeIDRef_);
             IDManager::instance().setID(reasonCodeIDRef_,reasonCode_);
          }
          else if(reasonCodeNode->Attribute("href")) { reasonCodeIDRef_ = reasonCodeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { reasonCode_ = boost::shared_ptr<ReasonCode>(new ReasonCode(reasonCodeNode));}
   }

   //locationNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* locationNode = xmlNode->FirstChildElement("location");

   if(locationNode){locationIsNull_ = false;}
   else{locationIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- locationNode , address : " << locationNode << std::endl;
   #endif
   if(locationNode)
   {
      if(locationNode->Attribute("href") || locationNode->Attribute("id"))
      {
          if(locationNode->Attribute("id"))
          {
             locationIDRef_ = locationNode->Attribute("id");
             location_ = boost::shared_ptr<ProblemLocation>(new ProblemLocation(locationNode));
             location_->setID(locationIDRef_);
             IDManager::instance().setID(locationIDRef_,location_);
          }
          else if(locationNode->Attribute("href")) { locationIDRef_ = locationNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { location_ = boost::shared_ptr<ProblemLocation>(new ProblemLocation(locationNode));}
   }

   //descriptionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* descriptionNode = xmlNode->FirstChildElement("description");

   if(descriptionNode){descriptionIsNull_ = false;}
   else{descriptionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- descriptionNode , address : " << descriptionNode << std::endl;
   #endif
   if(descriptionNode)
   {
      if(descriptionNode->Attribute("href") || descriptionNode->Attribute("id"))
      {
          if(descriptionNode->Attribute("id"))
          {
             descriptionIDRef_ = descriptionNode->Attribute("id");
             description_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(descriptionNode));
             description_->setID(descriptionIDRef_);
             IDManager::instance().setID(descriptionIDRef_,description_);
          }
          else if(descriptionNode->Attribute("href")) { descriptionIDRef_ = descriptionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { description_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(descriptionNode));}
   }

   //validationRuleIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* validationRuleIdNode = xmlNode->FirstChildElement("validationRuleId");

   if(validationRuleIdNode){validationRuleIdIsNull_ = false;}
   else{validationRuleIdIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- validationRuleIdNode , address : " << validationRuleIdNode << std::endl;
   #endif
   if(validationRuleIdNode)
   {
      if(validationRuleIdNode->Attribute("href") || validationRuleIdNode->Attribute("id"))
      {
          if(validationRuleIdNode->Attribute("id"))
          {
             validationRuleIdIDRef_ = validationRuleIdNode->Attribute("id");
             validationRuleId_ = boost::shared_ptr<Validation>(new Validation(validationRuleIdNode));
             validationRuleId_->setID(validationRuleIdIDRef_);
             IDManager::instance().setID(validationRuleIdIDRef_,validationRuleId_);
          }
          else if(validationRuleIdNode->Attribute("href")) { validationRuleIdIDRef_ = validationRuleIdNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { validationRuleId_ = boost::shared_ptr<Validation>(new Validation(validationRuleIdNode));}
   }

   //additionalDataNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* additionalDataNode = xmlNode->FirstChildElement("additionalData");

   if(additionalDataNode){additionalDataIsNull_ = false;}
   else{additionalDataIsNull_ = true;}

   if(additionalDataNode)
   {
      for(additionalDataNode; additionalDataNode; additionalDataNode = additionalDataNode->NextSiblingElement("additionalData")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- additionalDataNode , address : " << additionalDataNode << std::endl;
          #endif
          if(additionalDataNode->Attribute("href") || additionalDataNode->Attribute("id"))
          {
              if(additionalDataNode->Attribute("id"))
              {
                  additionalDataIDRef_ = additionalDataNode->Attribute("id");
                  additionalData_.push_back(boost::shared_ptr<AdditionalData>(new AdditionalData(additionalDataNode)));
                  additionalData_.back()->setID(additionalDataIDRef_);
                  IDManager::instance().setID(additionalDataIDRef_, additionalData_.back());
              }
              else if(additionalDataNode->Attribute("href")) { additionalDataIDRef_ = additionalDataNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { additionalData_.push_back(boost::shared_ptr<AdditionalData>(new AdditionalData(additionalDataNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- additionalDataNode , address : " << additionalDataNode << std::endl;
       #endif
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<ReasonCode> Reason::getReasonCode()
{
   if(!this->reasonCodeIsNull_){
        if(reasonCodeIDRef_ != ""){
             return boost::shared_static_cast<ReasonCode>(IDManager::instance().getID(reasonCodeIDRef_));
        }else{
             return this->reasonCode_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ReasonCode>();
   }
}
boost::shared_ptr<ProblemLocation> Reason::getLocation()
{
   if(!this->locationIsNull_){
        if(locationIDRef_ != ""){
             return boost::shared_static_cast<ProblemLocation>(IDManager::instance().getID(locationIDRef_));
        }else{
             return this->location_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ProblemLocation>();
   }
}
boost::shared_ptr<XsdTypeString> Reason::getDescription()
{
   if(!this->descriptionIsNull_){
        if(descriptionIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeString>(IDManager::instance().getID(descriptionIDRef_));
        }else{
             return this->description_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeString>();
   }
}
boost::shared_ptr<Validation> Reason::getValidationRuleId()
{
   if(!this->validationRuleIdIsNull_){
        if(validationRuleIdIDRef_ != ""){
             return boost::shared_static_cast<Validation>(IDManager::instance().getID(validationRuleIdIDRef_));
        }else{
             return this->validationRuleId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Validation>();
   }
}
std::vector<boost::shared_ptr<AdditionalData>> Reason::getAdditionalData()
{
   if(!this->additionalDataIsNull_){
        if(additionalDataIDRef_ != ""){
             return this->additionalData_;
        }else{
             return this->additionalData_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<AdditionalData>>();
   }
}
}

